File: DWM/pagerank.java
package DWM;

import java.util.Arrays;
public class pagerank{
    public static void main(String[] args) {
        //step 1 create link matrix, which page is linked to what
        double[][] link = {
                {0, 1, 0, 1},
                {1, 0, 0, 1},
                {1, 1, 0, 0},
                {0, 0, 1, 0}
        };

        //create variables
        int n = link.length;
        double dampingFactor = 0.85;
        double tolerance = 0.0001;

        //create initial pagerank matrix and initialize it with 1/n
        double[] pageRank = new double[n];
        Arrays.fill(pageRank, 1.0/n);

        //formula is
        // (1-d)/n + d*sumation of (pageRank/outgoinglink)

        //so first we will calculate outgoing link for each page
        double[] outGoing = new double[n];
        for (int i = 0; i < n; i++) {
            double sum=0;
            for (int j = 0; j < n; j++) {
                sum+=link[i][j];
            }
            outGoing[i] = sum;
        }

        //now we will calculate actual pagerank until convergence
        boolean convergence = false;
        int iterartion = 0;

        while (!convergence){
            double[] newPageRank = new double[n];
            for (int i = 0; i <n; i++) {
                double sum = 0;
                for (int j = 0; j < n; j++) {
                    if (link[i][j] == 1){
                        sum += pageRank[j]/outGoing[j];
                    }  
                }
                newPageRank[i] = (1-dampingFactor)/n + dampingFactor*sum;
            }

            convergence = true;
            for (int i = 0; i < n; i++) {
                if(Math.abs(newPageRank[i] - pageRank[i]) > tolerance){
                    convergence = false;
                    break;
                }
            }
            pageRank = newPageRank;
            iterartion++;
        }

        System.out.println("pagerank after " + iterartion + " are");
        for (int i = 0; i < n; i++) {
            System.out.println("page-" + (i+1) + ": " + pageRank[i]);
        }
    }
}

File: DWM/pagerank.py
# Step 1: Create link matrix (which page is linked to what)
link = [
    [0, 1, 0, 1],
    [1, 0, 0, 1],
    [1, 1, 0, 0],
    [0, 0, 1, 0]
]

# Create variables
n = len(link)
damping_factor = 0.85
tolerance = 0.0001

# Create initial PageRank matrix and initialize with 1/n
page_rank = [1.0 / n] * n

# Step 2: Calculate outgoing links for each page
out_going = []
for i in range(n):
    sum_links = 0
    for j in range(n):
        sum_links += link[i][j]
    out_going.append(sum_links)

# Step 3: Calculate PageRank until convergence
convergence = False
iteration = 0

while not convergence:
    new_page_rank = [0.0] * n
    for i in range(n):
        sum_val = 0
        for j in range(n):
            if link[i][j] == 1:
                sum_val += page_rank[j] / out_going[j]
        new_page_rank[i] = (1 - damping_factor) / n + damping_factor * sum_val

    convergence = True
    for i in range(n):
        if abs(new_page_rank[i] - page_rank[i]) > tolerance:
            convergence = False
            break

    page_rank = new_page_rank
    iteration += 1

# Step 4: Print result
print("PageRank after", iteration, "iterations:")
for i in range(n):
    print("Page", i + 1, ":", page_rank[i])